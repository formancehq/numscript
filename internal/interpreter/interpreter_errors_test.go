package interpreter_test

import (
	"testing"

	"github.com/formancehq/numscript/internal/interpreter"
	"github.com/formancehq/numscript/internal/parser"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/require"
)

func matchErrWithSnapshots(t *testing.T, src string, runOpt interpreter.RunProgramOptions) {
	parsed := parser.Parse(src)
	_, err := interpreter.RunProgram(parsed.Value, runOpt)
	require.NotNil(t, err)
	snaps.MatchSnapshot(t, err.GetRange().ShowOnSource(parsed.Source))
}

func TestShowUnboundVar(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = $unbound_var
  destination = @dest
)`, interpreter.RunProgramOptions{})

}

func TestShowMissingFundsSingleAccount(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = @a
  destination = @dest
)`, interpreter.RunProgramOptions{})

}

func TestShowMissingFundsInorder(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = {
    @a
    @b
	}
  destination = @dest
)`, interpreter.RunProgramOptions{})
}

func TestShowMissingFundsAllotment(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = {
    1/2 from @a
		remaining from @world
	}
  destination = @dest
)`, interpreter.RunProgramOptions{})
}

func TestShowMissingFundsMax(t *testing.T) {
	matchErrWithSnapshots(t, `send [COIN 10] (
  source = max [COIN 2] from {
    1/2 from @world
    remaining from @world
  }
  destination = @dest
)`, interpreter.RunProgramOptions{})
}

func TestShowMetadataNotFound(t *testing.T) {
	matchErrWithSnapshots(t, `vars {
  number $my_var = meta(@acc, "key")
}
`, interpreter.RunProgramOptions{})
}

func TestShowTypeError(t *testing.T) {
	matchErrWithSnapshots(t, `send 42 (
	source = @a
	destination = @b
)
`, interpreter.RunProgramOptions{})
}

func TestShowInvalidTypeErr(t *testing.T) {
	matchErrWithSnapshots(t, `vars {
  invalid_t $x
}
`, interpreter.RunProgramOptions{
		Vars: map[string]string{"x": "42"},
	})
}
