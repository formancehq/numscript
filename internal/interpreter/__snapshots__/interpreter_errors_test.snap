
[TestShowUnboundVar - 1]
  0 | send [COIN 10] (
  1 |   source = $unbound_var
    |            ~~~~~~~~~~~~
  2 |   destination = @dest

---

[TestShowMissingFundsSingleAccount - 1]
  0 | send [COIN 10] (
  1 |   source = @a
    |            ~~
  2 |   destination = @dest

---

[TestShowMissingFundsInorder - 1]
  0 | send [COIN 10] (
  1 |   source = {
    |            ~
  2 |     @a
    | ~~~~~~
  3 |     @b
    | ~~~~~~
  4 |  }
    | ~~
  5 |   destination = @dest

---

[TestShowMissingFundsAllotment - 1]
  1 |   source = {
  2 |     1/2 from @a
    |              ~~
  3 |      remaining from @world

---

[TestShowMissingFundsMax - 1]
  0 | send [COIN 10] (
  1 |   source = max [COIN 2] from {
    |            ~~~~~~~~~~~~~~~~~~~
  2 |     1/2 from @world
    | ~~~~~~~~~~~~~~~~~~~
  3 |     remaining from @world
    | ~~~~~~~~~~~~~~~~~~~~~~~~~
  4 |   }
    | ~~~
  5 |   destination = @dest

---

[TestShowMetadataNotFound - 1]
  0 | vars {
  1 |   number $my_var = meta(@acc, "key")
    |                    ~~~~~~~~~~~~~~~~~
  2 | }

---

[TestShowTypeError - 1]
  0 | send 42 (
    |      ~~
  1 |  source = @a

---

[TestShowInvalidTypeErr - 1]
  0 | vars {
  1 |   invalid_t $x
    |   ~~~~~~~~~
  2 | }

---
